### PR TITLE
Support tuple

### PR DESCRIPTION
**Pull request by phlptp**
If merged this pull request will enable full support for tuples up to 5 elements in size.   This includes support for std::pair and std::array(up to 5 elements).

### CODE REVIEW

**Comment by phlptp** [comment]
@henryiii   this is something I have been tinkering with and I wanted to see how close I am on coverage and platform support.  There is a bit more work to be done on testing, and possible support for things like vector<pair> 

I looked at the complex,  but there is a special support for "j" notation on complex numbers so can't get rid of that yet. 

you can take a look see if this is of interest

**Comment by phlptp** [comment]
ok, obviously have a little ways to go to get broad compiler support

**Comment by phlptp** [comment]
This should address #296 since you can now do
```cpp
std::pair<size_t, int> index;
app.add_option("--entry",index,"pair option");  
```

File: include/CLI/TypeTools.hpp [hunk 1]
```diff
@@ -390,6 +448,60 @@ constexpr const char *type_name() {
...
+    typename T,
+    enable_if_t<classify_object<T>::value == tuple_value && type_count<T>::value == 4, detail::enabler> = detail::dummy>
+std::string type_name() {
+    return std::string("[") + type_name<typename std::tuple_element<0, T>::type>() + "," +
+           type_name<typename std::tuple_element<1, T>::type>() + "," +
+           type_name<typename std::tuple_element<2, T>::type>() + "," +
+           type_name<typename std::tuple_element<3, T>::type>() + "]";
+}
+
+/// Print type name for 5 element tuples
```
Comment (henryiii) [inline include/CLI/TypeTools.hpp:494]: Is the limit of 5 here just the printout for the typename?

File: include/CLI/TypeTools.hpp [hunk 2]
```diff
@@ -607,6 +719,160 @@ bool lexical_assign(const std::string &input, T &output) {
...
+        retval &= lexical_cast(strings[2], std::get<2>(output));
+    }
+    if(strings.size() > 3) {
+        retval &= lexical_cast(strings[3], std::get<3>(output));
+    }
+    return retval;
+}
+
+/// conversion for five element tuple
+template <class T, class XC, enable_if_t<type_count<T>::value == 5, detail::enabler> = detail::dummy>
```
Comment (henryiii) [inline include/CLI/TypeTools.hpp:856]: It seems like there should be a recursive way to do this.

File: include/CLI/TypeTools.hpp (hunk 1 above)
Comment (phlptp) [reply include/CLI/TypeTools.hpp:494]: It is all the places tuples are supported.  I tried to figure out how to do that with some template magic, but the initial solutions I tried got out of hand quickly so I didn't really like that solution.  On my other projects I had access to Boost so using the MPL  library or one of the other TMP libraries this gets all pretty easy, but I am a little scared to try to look under the hood for that stuff, and it didn't seem appropriate to start including Boost now.  So 5 is one greater than the biggest  tuple case I could conceivably find a use case for, which is where that came from.

File: include/CLI/TypeTools.hpp (hunk 2 above)
Comment (phlptp) [reply include/CLI/TypeTools.hpp:856]: For some parts I think it is doable with a compile time recursion, but it wasn't clear to me how to transmit the runtime state along with the compile time recursion and keep everything sane.  Would love it if you or anyone knew how to do that but I was not seeing any clear solutions to get around the limit of 5 in a nice way.  Not sure how much value there is in supporting 6 element tuples either.

**Comment by henryiii** [comment]
mp11 or (probably) hanna can do it, so it should be possible. For now, this is probably fine and is a nice feature to have.

**Comment by phlptp** [comment]
I think my real question is not whether it can be done, but whether it can be done a straightforward and somewhat readable way.

**Comment by phlptp** [comment]
From #296 what I don't know quite yet is how to handle the validators that might be distinct for each part of a tuple.   Not sure the best way to handle that yet.

**Comment by henryiii** [comment]
A future improvement could be looking at how to do this without a limit and 5x repeated code.

**Pull request merged**